### PR TITLE
Modify webcontainer LTC management to support transaction config updates

### PR DESCRIPTION
**Pull request by dazavala**
When the transaction configuration and a transactional data source configuration are simultaneously updated, either by the user or by checkpoint-restore config overrides, the web container will incorrectly manage LTCs when dispatching a servlet request.  The failure occurs because the web container does not use the updated WebAppTransactionsCollaborator service instance that registers during the transaction config update, but rather, it uses a stale instance which no longer references the services necessary to manage LTCs.

This PR provides the following:

1) A simple test, `MockServletTest`, that exercises a server with stubbed CRIU operation. A stubbed server runs checkpoint-restore in a single jvm and affords the attachment of a debugger to examine server operation.

2) A fix to the WebContainer CollaboratorHelper that obtains the currently registered instance of the WebAppTransactionCollaborator service before exercising preInvoke and postInvoke behaviors on a servlet dispatch.

3) A fix to the transaction configuration service that enables the transactionLogDir setting to update at restore. Note: this commit also reverts a nit I introduced in the solution to defer recovery during restore.

4) Modifications to the checkpoint transaction FATs for Servlets (UserTransaction) that update the transaction and data source configurations at restore and verify server behavior.  These changes verify fixes 2) and 3) by invoking servlet methods that require LTC management and that update the tranlog path at restore.

5) Modifications to the checkpoint transaction FATs for EJBs (injection, BMT, CMT), ManagedBeans (injection, Transactional, TransactionScoped) and Recovery that update the transaction configuration at restore and verify server behavior. These changes discovered a regression in the CDI feature, which failed to inject managed bean reference after config update -- see fix 6).

6) Fixes to CDI service components, `CDIRuntime` and `LibertyTransactionService`, that declare references to the `UserTransaction` and `TransactionManagerService` services as `DYNAMIC`+`OPTIONAL`. These metadata changes enable the declarative services runtime to inject updated service references w/o reactivating the CDI components.

7) A new test, `Simple2PCCloudTest`, that exercises simple peer recovery scenarios involving three servers that log transactions to a shared datasource.  Two of the servers do checkpoint and restore with datasource and transaction configuration updates, including `recoveryIdentity`s containing both specified and built-in values for ${HOSTNAME}.

### CODE REVIEW

**Comment by dazavala** [comment]
#build
Let's get build results before review.

**Comment by dazavala** [comment]
@jonhawkes Please have a look at the TM changes. I can reproduce this break on a plain ol' server. Additional tests in the TX FATs may be warranted. Also, after we deliver these changes, **I intend to retrofit some peer recovery tests for InstantOn. Would you point me to the most essential FATs and/or test scenarios?**   

@pmd1nh Please have a look at the web container changes. I maintained the consistency of your code.

@tjwatson The commits pushed tonight are nearly identical to what you viewed earlier and include your suggestions.  Going forward I think we should add tran config updates to some container-managed and transactional beans scenarios within the checkpoint FAT.

File: dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/collaborator/CollaboratorHelperImpl.java [hunk 1]
```diff
@@ -38,6 +38,7 @@
 import com.ibm.ws.runtime.metadata.ComponentMetaData;
 import com.ibm.ws.threadContext.ComponentMetaDataAccessorImpl;
 import com.ibm.ws.webcontainer.collaborator.WebAppSecurityCollaborator;
+import com.ibm.ws.webcontainer.collaborator.WebAppTransactionCollaborator;
 import com.ibm.ws.webcontainer.webapp.WebAppDispatcherContext;
```
Comment (pmd1nh) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/collaborator/CollaboratorHelperImpl.java:42]: update the Copyright year

Comment (dazavala) [reply dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/collaborator/CollaboratorHelperImpl.java:42]: @pmd1nh Done and thank you. See commit labeled "Review changes".

**Review by pmd1nh — DISMISSED** [review]
WC change looks good.  Thanks!

**Comment by dazavala** [comment]
Hey @benjamin-confino (fyi: @tjwatson @jonhawkes). Please review commit "[CDI changes for transaction config updates](https://github.com/OpenLiberty/open-liberty/pull/25430/commits/194284f3c8f47cfdc434483a4788801d0821cc6d)". 

I added the `optional` cardinality to service references related to transaction support. The references are now `dynamic`+`optional`, which enables the declarative services runtime to inject updated service references w/o reactivating the `CDIRuntime` and `LibertyTransactionService (LTS)` service components.  Accessor methods for these services in classes `LTS` and `CDIRuntimeImpl` support dynamic updates to transaction-related services and require no change. I would prefer the classes maintain `volatile` references to reinforce they are dynamic rather then static. Scanning the CDI feature code, I found a suspicious path that caches a `TransactionService` reference. I see no problem here -- the referenced object is an `LTS`. Please advise if I missed something and thanks for your assistance thus far.

>     public WebSphereCDIDeploymentImpl(Application application, CDIRuntime cdiRuntime) {
>         this.bootstrap = new WeldBootstrap();
> 
>         this.id = application.getJ2EEName().toString();
>         this.application = application;
> 
>         this.serviceRegistry = new SimpleServiceRegistry();
>         TransactionService wsTransactionService = cdiRuntime.getTransactionService();
>         TransactionServicesImpl transactionServices = null;
>         if (wsTransactionService != null) {
>             transactionServices = new TransactionServicesImpl(wsTransactionService);
>         ... ... ...

**Comment by dazavala** [comment]
#build
Verify the initial changes to CDI.

**Comment by dazavala** [comment]
@sebratton Would you please review the following commits, which modify the checkpoint transaction FAT.  I'm mostly concerned about the new tests involving stubbed criu operation (`MockServletTest`) and the peer recovery test (`Simple2PCCoudTest`).  The other changes are simple modifications to flush out bugs caused by multiple configuration changes at restore.

[Add MockServletTest ...](https://github.com/OpenLiberty/open-liberty/pull/25430/commits/cef4005b23daa3b8f1fc99889fd4bcb7529236e0)
[Verify simple transaction config updates...](https://github.com/OpenLiberty/open-liberty/pull/25430/commits/d30f6f4721440f8cf0dcfa6bc12b4f0c6bddeb1d)
[Modify EJB, managedbean, ...](https://github.com/OpenLiberty/open-liberty/pull/25430/commits/e319e53a43c75a9b5c92d4e4f521dcc43fc5b081)
[Add peer recovery test ...](https://github.com/OpenLiberty/open-liberty/pull/25430/commits/5522396fe2e06240b8fda7951ddb0ff73de9bdde)

**Comment by dazavala** [comment]
#build

**Review by sebratton — DISMISSED** [review]

**Comment by dazavala** [comment]
#libby

**Comment by dazavala** [comment]
#build 
Again

**Comment by dazavala** [comment]
#build 
Add `@SkipIfCheckpointNotSupported` to new test `SimplePCCloudTest.java`.

**Comment by dazavala** [comment]
#libby

**Review by sebratton — APPROVED** [review]

**Comment by dazavala** [comment]
#libby

**Pull request merged**